src/Interfaces/FwupdManager.vala

```diff
@@ -0,0 +1,302 @@
+                    break;
+                case "RemoteId":
+                    release.remote_id = val.get_string ();
+                    break;
+                case "AppstreamId":
+                    release.appstream_id = val.get_string ();
+                    break;
+                case "Checksum":
+                    release.checksum = val.get_string ();
+                    break;
+                case "Vendor":
+                    release.vendor = val.get_string ();
+                    break;
+                case "Size":
+                    release.size = val.get_uint64 ();
+                    break;
+                case "License":
+                    release.license = val.get_string ();
+                    break;
+                case "TrustFlags":
+                    release.flag = (Fwupd.ReleaseFlag) val.get_uint64 ();
+                    break;
+                case "InstallDuration":
+                    release.install_duration = val.get_uint32 ();
+                    break;
+                case "Uri":
+                    release.uri = val.get_string ();
+                    break;
+                default:
+                    break;
+            }
+        });
+
+        return release;
+    }
+
+    private string get_path (string uri) {
+        var parts = uri.split ("/");
+        string file_path = parts[parts.length - 1];
+        return Path.build_path (Path.DIR_SEPARATOR_S, Environment.get_tmp_dir (), file_path);
+    }
+
+    public async string? download_file (Fwupd.Device device, string uri) {
+        var path = get_path (uri);
+
+        File server_file = File.new_for_uri (uri);
+        File local_file = File.new_for_path (path);
+
+        bool result;
+        try {
+            result = yield server_file.copy_async (local_file, FileCopyFlags.OVERWRITE, Priority.DEFAULT, null, (current_num_bytes, total_num_bytes) => {});
+        } catch (Error e) {
+            on_device_error (device, "Could not download file: %s".printf (e.message));
+            return null;
+        }
+
+        if (!result) {
+            on_device_error (device, "Download of %s was not succesfull".printf (uri));
+            return null;
+        }
+
+        return path;
+    }
+
+    public async bool install (Fwupd.Device device, string path) {
+        try {
+            // https://github.com/fwupd/fwupd/blob/c0d4c09a02a40167e9de57f82c0033bb92e24167/libfwupd/fwupd-client.c#L2045
+            var options = new GLib.HashTable<string, Variant> (str_hash, str_equal);
+            options.insert ("reason", new Variant.string ("user-action"));
+            options.insert ("filename", new Variant.string (path));
+            options.insert ("allow-older", new Variant.boolean (true));
+            options.insert ("allow-reinstall", new Variant.boolean (true));
+            options.insert ("no-history", new Variant.boolean (true));
+
+            var fd = Posix.open (path, Posix.O_RDONLY);
```

Comment:
Are you supposed to close this file descriptor once done, or does fwupd handle that?

Reply:
To test this I called `Posix.close` twice. The first time it returned `0`. The second time it returned `-1`. So I suppose we have to take care of closing it ourselves.

I refactored the code and added `Posix.close (fd);` in a `finally` statement.